frontend/src/metabase/dashboard/components/DashboardEmbedAction/DashboardEmbedAction.tsx

```diff
@@ -0,0 +1,116 @@
+import { useState } from "react";
+import type { Dashboard } from "metabase-types/api";
+import { DashboardEmbedHeaderButton } from "metabase/dashboard/components/DashboardEmbedHeaderButton";
+import { DashboardEmbedHeaderMenu } from "metabase/dashboard/components/DashboardEmbedHeaderMenu";
+import DashboardSharingEmbeddingModal from "metabase/dashboard/containers/DashboardSharingEmbeddingModal";
+import { useSelector } from "metabase/lib/redux";
+import { getSetting } from "metabase/selectors/settings";
+import { getUserIsAdmin } from "metabase/selectors/user";
+
+export type EmbedButtonClickBehavior =
+  | "embed-menu"
+  | "embed-modal"
+  | "public-link-popover"
+  | null;
+
+const getClickBehavior = ({
+  isAdmin,
+  isPublicSharingEnabled,
+  hasPublicLink,
+}: {
+  isAdmin: boolean;
+  isPublicSharingEnabled: boolean;
+  hasPublicLink: boolean;
+}): EmbedButtonClickBehavior => {
+  if (!isPublicSharingEnabled) {
+    return isAdmin ? "embed-modal" : null;
+  }
+
+  if (isAdmin) {
+    return "embed-menu";
+  }
+
+  if (hasPublicLink) {
+    return "public-link-popover";
+  }
+
+  return null;
```

Comment:
Maybe we could simplify this condition? I came up with this solution, maybe you could verify if it's still working correctly. The current condition is a bit hard to follow because we're returning `null` in 2 places:
1. `!isPublicSharingEnabled && !isAdmin`
2. `isPublicSharingEnabled && !isAdmin && !hasPublicLink`

```suggestion
  if (isAdmin) {
    return isPublicSharingEnabled ? "embed-menu" : "embed-modal";
  }

  if (isPublicSharingEnabled && hasPublicLink) {
    return "public-link-popover";
  }

  return null;
```

Reply:
Hahahaha I spent a while on this one, since we have to play with the `isAdmin`, `isPublicSharing`, and the `null` value. My first thought was to ensure that public sharing would be the condition that's only mentioned once, only because there's three branches that depend on that condition. But I like this as well.